File: sdk/android/tokenpay-id-sdk/src/main/java/space/tokenpay/id/TpidResult.kt
package space.tokenpay.id

/**
 * Result of a widget session.
 */
sealed class TpidResult {
    /** User authenticated successfully. */
    data class Success(
        val accessToken: String,
        val refreshToken: String?,
        val idToken: String?,
        val expiresIn: Long,
        val tokenType: String = "Bearer",
        val scope: String,
        val user: TpidUser,
    ) : TpidResult()

    /** User dismissed the widget without completing auth. */
    data object Cancelled : TpidResult()

    /** Widget finished with an error. Inspect [error] for the reason. */
    data class Failure(val error: TpidError) : TpidResult()
}

/**
 * Authenticated user profile returned alongside tokens.
 */
data class TpidUser(
    val id: String,
    val email: String,
    val emailVerified: Boolean,
    val name: String?,
    val avatarUrl: String?,
    val locale: String?,
    val has2fa: Boolean,
    val hasPasskey: Boolean,
    val createdAt: String,
)

File: sdk/swift/Sources/TokenPayID/TpidResult.swift
import Foundation

/// Result of a widget session.
public enum TpidResult: Sendable {
    case success(TpidSession)
    case cancelled
    case failure(TpidError)
}

public struct TpidSession: Sendable, Codable, Equatable {
    public let accessToken: String
    public let refreshToken: String?
    public let idToken: String?
    public let expiresAt: Date
    public let tokenType: String
    public let scope: String
    public let user: TpidUser

    public var isValid: Bool { expiresAt.timeIntervalSinceNow > 30 }
}

public struct TpidUser: Sendable, Codable, Equatable {
    public let id: String
    public let email: String
    public let emailVerified: Bool
    public let name: String?
    public let avatarURL: URL?
    public let locale: String?
    public let has2FA: Bool
    public let hasPasskey: Bool
    public let createdAt: String

    public init(
        id: String, email: String, emailVerified: Bool,
        name: String? = nil, avatarURL: URL? = nil, locale: String? = nil,
        has2FA: Bool = false, hasPasskey: Bool = false, createdAt: String = ""
    ) {
        self.id = id; self.email = email; self.emailVerified = emailVerified
        self.name = name; self.avatarURL = avatarURL; self.locale = locale
        self.has2FA = has2FA; self.hasPasskey = hasPasskey; self.createdAt = createdAt
    }
}
